src/crawl_client.rs

```diff
@@ -0,0 +1,66 @@
+use crate::{
+	network::Client,
+	telemetry::{MetricValue, Metrics},
+	types::{self, delay_for},
+};
+use avail_subxt::primitives::Header;
+use kate_recovery::matrix::Partition;
+use std::{
+	sync::Arc,
+	time::{Duration, Instant, SystemTime},
+};
+use tokio::sync::broadcast;
+use tracing::{error, info};
+
+const ENTIRE_BLOCK: Partition = Partition {
+	number: 1,
+	fraction: 1,
+};
+
+pub async fn run(
+	mut message_rx: broadcast::Receiver<(Header, Instant)>,
+	network_client: Client,
+	delay: u64,
+	metrics: Arc<impl Metrics>,
+) {
+	info!("Starting crawl client...");
+
+	while let Ok((header, received_at)) = message_rx.recv().await {
+		if let Some(seconds) = delay_for(Duration::from_secs(delay), received_at.elapsed()) {
+			info!("Sleeping for {seconds:?} seconds");
+			tokio::time::sleep(seconds).await;
+		}
+		let block_number = header.number;
+		info!("Crawling block {block_number}...");
+
+		let start = SystemTime::now();
+
+		let Ok(block) = types::BlockVerified::try_from(header) else {
+			error!("Header is not valid");
+			continue;
+		};
```

Comment:
We can report an error here

Reply:
We can also move this check before the sleep, so we won't waste queue for not valid headers

Reply:
I have moved it before the sleep, but there is no significant change since sleep is used to postpone each block processing for at least given amount of second, and each block is produced in regular intervals (e.g. 20 seconds).